includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,356 @@
  * @param AbstractObject $object An Islandora Object
  * @param String $filePath video file path
  */
+
+
+libraries_load('aws'); 
+
+use Aws\S3\S3Client;
+use Aws\Exception\AwsException;
+use Aws\S3\ObjectUploader;
+use Aws\S3\MultipartUploader;
+use Aws\Exception\MultipartUploadException;
+
+# Filename of manifest XML file. Any filename is acceptable.
+const MANIFEST_FILE_NAME = '/tmp/upload_manifest_generated.xml';
+
 function islandora_video_streaming_panopto_upload($object, $filePath) {
-  if ($force || !isset[$object['KALTURA_STREAM']) {
+  if ($filePath || !isset($object['PANOPTO_INFO'])) {
+    module_load_include('inc', 'islandora_video', "includes/derivatives" );
     $archival_file = islandora_video_copy_obj_datastream($object);
     if ($archival_file === FALSE) {
       return islandora_video_no_obj_datastream($object->id);
     }
     $filePath = drupal_realpath($archival_file['file']->uri);
     $failureMessage = array(
       'success' => false,
-      'messages' => array(
-        'message' => t('Failed to upload object @pid to Panopto, but no exception caught.'),
+      'messages' => array(),
+    );
+    $streaming = array();
+    
+    $oauthendpoint = variable_get('islandora_video_streaming_panopto_host');
+    $uploadendpoint = $oauthendpoint . '/Panopto/PublicAPI/Rest/sessionUpload';
+    $oauthendpoint .= "/Panopto/oauth2/connect/token";
+    $token = islandora_video_streaming_panopto_get_token($oauthendpoint);
+    $session = null;
+    if ($token) {
+      $session = islandora_video_streaming_panopto_get_session($token, $uploadendpoint); 
+    }
+    else {
+      $messages = array(
+        'message' => t('Failed to retrieve token for object: @pid'),
         'message_sub' => array('@pid' => $object->id),
         'type' => 'watchdog',
         'severity' => WATCHDOG_ERROR,
-      ),
-    );
-
-    //TODO: upload, record result in $streaming
- 
+      );
+      $failureMessage['messages'][] = $messages;
+      return $failureMessage;
+    }
+    $success = false;
+    if ($session) {
+      try {
+        $upload_id = $session['ID'];
+        $upload_target = $session['UploadTarget'];
+        $file_uri = explode("/", $filePath);
+        $FILENAME = end($file_uri);
+        
+        islandora_video_streaming_panopto_create_manifest_for_video($filePath, MANIFEST_FILE_NAME);
```

Comment:
Constant `MANIFEST_FILE_NAME` is really only used on these two lines: passed in to `islandora_video_streaming_panopto_create_manifest_for_video` and into `islandora_video_streaming_panopto_upload_file`.  There is no need for it to be a global constant.  It should probably only exist in `islandora_video_streaming_panopto_create_manifest_for_video`.   `islandora_video_streaming_panopto_create_manifest_for_video` should return the name of the temporary file created, to be passed in to `islandora_video_streaming_panopto_upload_file`.  It also needs to be generated dynamically, based on a Drupal temp file.  If two users were to call this function at the same time, there would be a race condition to write to the same file.

Reply:
Done, but will need to check if any errors arise. Apologies for all the "Done" comments, I was having trouble keeping track of which ones I had finished.